pkg/rhcos/openstack.go

```diff
@@ -25,7 +34,114 @@ func OpenStack(ctx context.Context) (string, error) {
 		return "", err
 	}
 
-	return base.ResolveReference(relOpenStack).String(), nil
+	// Attach uncompressed sha256 checksum to the URL
+	checkSuffix := "?sha256=" + meta.Images.OpenStack.UncompressedSHA256
+	baseURL := base.ResolveReference(relOpenStack).String() + checkSuffix
+
+	// Check that we have generated a valid URL
+	_, err = url.ParseRequestURI(baseURL)
+	if err != nil {
+		return "", err
+	}
+
+	return baseURL, nil
+}
+
+func getCacheDir() (string, error) {
+	// TODO(mfedosin): os.UserCacheDir was added in golang 1.11, but since we still use 1.10
+	// we can't use it directly. Uncomment this when we start using golang 1.11+ for building
+	// the binary.
+	//defaultCacheDir, err := os.UserCacheDir()
+
+	defaultCacheDir := configdir.New("openshift-installer", "images-cache").QueryCacheFolder()
+
+	defaultCacheDir.MkdirAll()
+
+	return defaultCacheDir.Path, nil
+}
+
+func validateSHA256Checksum(filepath, checksum string) error {
+	f, err := os.Open(filepath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return err
+	}
+
+	if checksum != fmt.Sprintf("%x", h.Sum(nil)) {
+		logrus.Fatal("Image sha256 checksum is invalid.")
+		defer os.Remove(filepath)
+		return errors.Errorf("image sha256 checksum is invalid")
+	}
+
+	return nil
+}
+
+// DownloadOpenStackImageFile obtains a file from a given URL, puts it in the cache
+// and returns the local file path
+func DownloadOpenStackImageFile(baseURL string) (string, error) {
```

Comment:
i would expect all this download to move to pkg/tfvars like for libvirt https://github.com/openshift/installer/blob/master/pkg/tfvars/libvirt/cache.go

This pkg/rhcos is to getting image location, and not downloading them, that more suited to tfvars.

Reply:
Done